### PR TITLE
Closes #2040: Create decoder for decoding ICO files.

### LINKED ISSUE #2040: Create ICO decoder
A bunch of favicons are still ICO files. We need to be able to decode them. Code is in Fennec. Either we add this functionality to `browser-icons` or we create a standalone `lib-ico` if we think that could be useful in other situations.

### PR DESCRIPTION
**Pull request by pocmo**
Okay, I've been working on this branch for too long and I need to finally get this off my plate.

This is a Kotlin port and simplified version for the decoder we have in Fennec:
https://searchfox.org/mozilla-central/source/mobile/android/base/java/org/mozilla/gecko/icons/decoders/ICODecoder.java

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

File: components/browser/icons/src/main/java/mozilla/components/browser/icons/decoder/ico/IconDirectoryEntry.kt [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+                return 1
+            }
+
+            if (bitsPerPixel > other.bitsPerPixel) {
+                return -1
+            }
+        }
+
+        // Otherwise, take the larger of the BPP values.
+        if (bitsPerPixel > other.bitsPerPixel) {
```
Comment (Amejia481) [inline components/browser/icons/src/main/java/mozilla/components/browser/icons/decoder/ico/IconDirectoryEntry.kt]: Can the `if` statements be converted to a `when` statement?

File: components/browser/icons/src/main/java/mozilla/components/browser/icons/decoder/IconDecoder.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+     * Decodes the given [data] into a a [Bitmap] or null.
+     *
+     * The caller provides a maximum size. This is useful for image formats that may decode into multiple images. The
+     * decoder can use this size to determine which [Bitmap] to return.
+     */
+    fun decode(data: ByteArray, targetSize: Int, maxSize: Int, maxScaleFactor: Float): Bitmap?
+
+    companion object {
+        @Suppress("MagicNumber")
+        enum class ImageMagicNumbers(var value: ByteArray) {
```
[the remark this answers is not in the document]
Comment (pocmo) [reply components/browser/icons/src/main/java/mozilla/components/browser/icons/decoder/IconDecoder.kt]: Heh, in this case the term "magic number" is actually used:

> This is a list of file signatures, data used to identify or verify the content of a file. Such signatures are also known as magic numbers. 
https://en.wikipedia.org/wiki/List_of_file_signatures

File: components/browser/icons/src/main/java/mozilla/components/browser/icons/decoder/ico/IconDirectoryEntry.kt (hunk 1 above)
Comment (pocmo) [reply components/browser/icons/src/main/java/mozilla/components/browser/icons/decoder/ico/IconDirectoryEntry.kt]: That's a great idea. This code was a 1:1 conversion from Fennec and a `when` statement should be much nicer to read! I'll update the PR.

Comment (Amejia481) [reply components/browser/icons/src/main/java/mozilla/components/browser/icons/decoder/ico/IconDirectoryEntry.kt]: 👍

File: components/browser/icons/src/main/java/mozilla/components/browser/icons/decoder/IconDecoder.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Amejia481) [reply components/browser/icons/src/main/java/mozilla/components/browser/icons/decoder/IconDecoder.kt]: Every day we learn something new :)

**Pull request merged**